src/NBIHandler.cpp

```diff
@@ -0,0 +1,757 @@
+}
+
+/**
+ * Precomputes discretized points in the beam and maps the intersections of the flux surfaces.
+ */
+void NBIHandler::PrecomputeBeamMapLUT() {
+    ir_lut.resize(static_cast<size_t>(n_beam_theta) * n_beam_radius * n_beam_s,
+                  static_cast<len_t>(-1));
+
+    for (len_t it = 0; it < n_beam_theta; ++it) {
+        for (len_t irad = 0; irad < n_beam_radius; ++irad) {
+            for (len_t is = 0; is < n_beam_s; ++is) {
+                real_t beam_theta = it * d_beam_theta;
+                real_t beam_radius = (irad) *d_beam_radius;
+                real_t beam_s = s_start + is * d_beam_s;
+
+                len_t ir_now =
+                    CalculatePencilBeamFindFlux(beam_s, beam_radius, beam_theta);
+                ir_lut[lut_index(it, irad, is)] = ir_now;
+            }
+        }
+    }
+}
+
+/**
+ * Destructor
+ */
+NBIHandler::~NBIHandler() {
+    delete[] NBIHeatTerm_e;
+    delete[] NBIHeatTerm_i;
+    delete[] Deposition_profile;
+    delete[] Deposition_profile_times_Vprime;
+    delete[] dV_beam_prime_tot;
+    delete[] H_r_dTe;
+    delete[] H_r_dne;
+    delete[] d_NBIHeatTerm_e_d_Te;
+    delete[] d_NBIHeatTerm_e_d_ne;
+    delete[] d_NBIHeatTerm_i_d_Te;
+    delete[] d_NBIHeatTerm_i_d_ne;
+    delete[] d_NBIHeatTerm_i_d_T_ij;
+    delete[] d_NBIHeatTerm_i_d_n_ij;
+    delete[] d_NBIHeatTerm_e_d_T_ij;
+    delete[] d_NBIHeatTerm_e_d_n_ij;
+    delete[] H_r_dn_ij;
+    delete[] H_r_dT_ij;
+}
+
+/**
+ * Build: Called at the start of every timestep.
+ * Compute added heating term for each timestep and flux surface radius
+ */
+void NBIHandler::Build(const real_t t, const real_t,
+    FVM::UnknownQuantityHandler* unknowns, real_t Ti_beam_input
+) {
+
+    // Setting values to zero since build is called multiple times per timestep
+    const real_t* VpVol = this->radialGrid->GetVpVol();
+    const real_t* dr = this->radialGrid->GetDr();
+    for (len_t ir = 0; ir < nr; ++ir) {
+        NBIHeatTerm_e[ir] = 0.0;
+        NBIHeatTerm_i[ir] = 0.0;
+        Deposition_profile[ir] = 0.0;
+        Deposition_profile_times_Vprime[ir] = 0.0;
+        dV_beam_prime_tot[ir] = 0.0;
+        H_r_dTe[ir] = 0.0;
+        H_r_dne[ir] = 0.0;
+        d_NBIHeatTerm_e_d_Te[ir] = 0.0;
+        d_NBIHeatTerm_e_d_ne[ir] = 0.0;
+        d_NBIHeatTerm_i_d_Te[ir] = 0.0;
+        d_NBIHeatTerm_i_d_ne[ir] = 0.0;
+        for (len_t iz = 0; iz < nZ; iz++) {
+            for (len_t Z0 = 0; Z0 <= ions->GetZ(iz); Z0++) {
+                H_r_dT_ij[idx(ir, iz, Z0)] = 0.0;
+                H_r_dn_ij[idx(ir, iz, Z0)] = 0.0;
+                d_NBIHeatTerm_i_d_n_ij[idx(ir, iz, Z0)] = 0.0;
+                d_NBIHeatTerm_i_d_T_ij[idx(ir, iz, Z0)] = 0.0;
+                d_NBIHeatTerm_e_d_n_ij[idx(ir, iz, Z0)] = 0.0;
+                d_NBIHeatTerm_e_d_T_ij[idx(ir, iz, Z0)] = 0.0;
+            }
+        }
+    }
+
+    // Set the beam power, either constant or time-dependent
+    real_t powerNow = 0;
+    if (this->Power_Profile != nullptr) {
```

Comment:
Why is it handled like this? Why can't the ``Power_Profile`` also be used for a time-constant power?